nupic/research/frameworks/vernon/mixins/noise_robustness_test.py

```diff
@@ -0,0 +1,136 @@
+
+
+class NoiseRobustnessTest:
+    """
+    Replaces the standard evaluate model function with a loop over the same function
+    that conducts a noise robustness test on each iteration. In the test, validation
+    accuracy is reported for samples with varying percentages of Gaussian
+    noise. The output of this mixin is a results dictionary which has all of the
+    standard metrics for the default evaluation loop and also reports the value of each
+    metric at each level of specified noise.
+    """
+
+    def setup_experiment(self, config):
+        """
+        :param config:
+            - noise_levels: Optional, a list of floats between 0 and 1 which specify
+            how many units of the input image will receive Gaussian noise. For
+            example, if the noise level is 0.1, then about 10% of the incoming
+            inputs per sample will receive additive Gaussian noise. Note that
+            the zero noise case is included by default, and leaving noise_levels out
+            of the config will result in 0.1 to 0.9 in increments of 0.1.
+            - noise_mean: The mean of the noise which will be added to the incoming
+            data, defaults to 0
+            - noise_std: The standard deviation of the noise which will be added to
+            the incoming data, defaults to 1
+
+        Example config:
+        config = dict(
+            noise_levels = [0.1, 0.25, 0.5, 0.9]
+            noise_mean = 0.2
+            noise_sd = 0.5
+        )
+
+        """
+        super().setup_experiment(config)
+        noise_levels = config.get("noise_levels", torch.arange(0.1, 1, 0.1))
+        assert all(
+            0 < noise <= 1 for noise in noise_levels
+        ), "Noise levels must be between (0, 1]"
+        noise_mean = config.get("noise_mean", 0)
+        noise_std = config.get("noise_std", 1.0)
+        default_evaluate_model_func = config.get("evaluate_model_func", evaluate_model)
+        self.evaluate_model = partial(
+            evaluate_model_with_noise,
+            evaluate_model_func=default_evaluate_model_func,
+            noise_levels=noise_levels,
+            noise_mean=noise_mean,
+            noise_std=noise_std,
+        )
+
+    @classmethod
+    def get_execution_order(cls):
+        eo = super().get_execution_order()
+        eo["setup_experiment"].append("NoiseRobustnessTest: Initialize")
+        return eo
+
+
+def evaluate_model_with_noise(
+    model,
+    loader,
+    device,
+    noise_levels=None,
+    noise_mean=0,
+    noise_std=1,
+    evaluate_model_func=None,
+    **kwargs,
+):
+    if noise_levels is None:
+        noise_levels = []
+    noise_results = {}
+    dataset_transform = loader.dataset.transform
+    zero_noise_results = evaluate_model_func(model, loader, device, **kwargs)
+    for noise_level in noise_levels:
+        loader.dataset.transform = transforms.Compose(
+            [dataset_transform, AddGaussianNoise(noise_level, noise_mean, noise_std)]
```

Comment:
instead of separately doing the "zero noise" version, you could do the following just for code niceness:
```
noise_levels.insert(0, 0.0)
for noise level in noise_levels:
    # Create transforms and eval model
```

Reply:
i originally had this, but when assigning to the results dict it overwrote mean_accuracy with mean_accuracy_0_noise so I just decided to do two separate runs

Reply:
I opted for a separate run with zero noise so that the results dict would have one copy of the original expected parameters, then the rest will have the same parameters with appended "_x_noise"